File: Python/77.Combinations.py
#写出n个数的中k个数的所有组合(C(n,k))
import copy
class Solution(object):
    def combine(self, n, k):
        self.nums , self.n , self.k = range(1,n+1) , n , k
        self.L = []
        self.dfs(0,0,[])
        return self.L

    def dfs(self,deep,start,tmp):
        if deep==self.k:
            t = copy.deepcopy(tmp)
            self.L.append(t)
            tmp.pop()
            return
        else:
            for i in range(start,self.n):
                tmp.append(self.nums[i])
                self.dfs(deep+1,i+1,tmp)
            if not tmp==[]:
                tmp.pop()


s = Solution()
print(s.combine(5,2))

File: C++/Backtracking/77.Combinations.cpp
/*
求出n个数中找出k个的所有组合。
*/
/*
思路：DFS递归。
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
	vector<vector<int>> result;
public:
    vector<vector<int>> combine(int n, int k) {
    	vector<int> tmp;
        helpCombine(tmp,1,n,k);
        return result;
    }
    void helpCombine(vector<int>& tmp,int start,int n,int k)
    {
    	if(k == 0)
    	{
    		result.push_back(tmp);
    		return;
    	}
    	for(int i=start; i <= n; ++i)
    	{
    		tmp.push_back(i);
    		helpCombine(tmp,i+1,n,k-1);
    		tmp.pop_back();
    	}
    }
};
int main()
{
	Solution s;
    vector<vector<int>> out = s.combine(4,3);
    for(auto vec:out)
    {
    	for(int n:vec)
    		cout<<n<<" ";
    	cout<<endl;
    }
}

